### PR TITLE
fix: keep inline comment authentication inside glab

### LINKED ISSUE #86: Remove plaintext GitLab token extraction from inline-comment workflows
## Security audit finding

ClawHub reports a **High / Unvalidated Output Injection** finding against the currently published `gitlab-cli-skills@1.13.25` package:

- Audit: https://clawhub.ai/vince-winkintel/skills/gitlab-cli-skills/security-audit
- Finding: the inline-comment example runs `glab config get token`, copies the stored long-lived GitLab credential into Python, and then places it in `PRIVATE-TOKEN` headers.
- Risk: the plaintext token can be exposed through process memory, logging, tracebacks, debugging, or future downstream changes even though the current example does not intentionally print it.
- Other audit signals are clean: VirusTotal reports 64/64 clean and ClawHub static analysis reports no suspicious patterns.

## Affected source and generated package

- `glab-mr/SKILL.md` — the “Always Use JSON Body” Python example extracts the token and uses raw `urllib` requests.
- `scripts/post-inline-comment.py` — the reusable helper repeats the same credential-extraction pattern in `get_token()` and sends raw `PRIVATE-TOKEN` requests.
- `scripts/README-inline-comments.md` and nearby guidance describe the helper as reading the token from the environment or glab config.
- `scripts/build-claude-skill.sh` merges `glab-mr/SKILL.md` into the published single-skill artifact, so the unsafe example reaches ClawHub and `claude-skill.zip`.

The original JSON-body design is intentional: form-encoded nested position fields can silently create a general MR note instead of an inline discussion. Current `glab api` supports `--input -`, allowing the workflow to preserve an exact JSON request body while keeping authentication inside `glab`.

## Required changes

1. Replace the raw Python token-extraction example with a `glab api --input -` workflow that:
   - verifies the intended GitLab hostname and visible account before a write;
   - fetches fresh MR version SHAs through `glab api`;
   - sends the nested discussion payload as literal JSON from stdin;
   - checks that the returned note has a non-null `position`;
   - never reads, copies, logs, or emits a stored token.
2. Refactor `scripts/post-inline-comment.py` so all authenticated API calls go through `glab api` using argument arrays and JSON stdin. Remove direct token discovery and `PRIVATE-TOKEN` header construction while preserving existing input validation, pagination, inline anchoring, line-code fallback, and output behavior.
3. Update the helper documentation to state that authentication remains inside `glab`; callers may configure glab normally, but the helper must not access token values.
4. Add regression tests using a fake `glab` executable/process boundary. Tests must cover GET/pagination, JSON POST stdin, non-inline response detection, and prove that source/generated artifacts contain no `glab config get token` credential-extraction example.
5. Rebuild and inspect the canonical merged `claude-skill.zip` artifact locally. Do not commit generated release artifacts unless repository policy already tracks them.
6. Harden terminal-output boundaries around the helper: do not echo raw comment bodies, reject control characters in printed file paths, and strip terminal control sequences plus sensitive headers from bounded `glab` error output.

## Acceptance criteria

- No executable example or helper path invokes `glab config get token` or manually constructs an authenticated `PRIVATE-TOKEN` request.
- `glab api --input -` preserves the JSON-body behavior required for nested GitLab MR discussion positions.
- The helper uses explicit host selection, validates the visible account/host before writes, and keeps secrets out of stdout/stderr.
- Raw comment bodies are not echoed, and user/API-controlled terminal control sequences cannot reach progress or error output.
- Existing input limits and line-code fallback behavior remain covered.
- Targeted regression tests and repository validation pass.
- The merged single-skill artifact no longer contains the ClawHub-reported token-extraction snippet.
- The pull request links this issue and documents validation plus any behavior/compatibility changes.

## Non-goals

- Do not publish a release or merge automatically.
- Do not broaden the change into unrelated GitLab CLI refresh work.
- Do not require or expose a real GitLab token in tests.

### PR DESCRIPTION
**Pull request by vince-winkintel**
## Summary

- remove plaintext GitLab token extraction and manual `PRIVATE-TOKEN` requests from inline-comment documentation and helper paths
- route authenticated GET/POST calls through `glab api`, using literal JSON on stdin for nested discussion positions
- verify the selected host/account before writes, preserve pagination and line-code fallback behavior, and harden terminal output/redaction
- sanitize API-controlled host, username, head SHA, and discussion ID output; reject C1 and bidirectional controls in printed file paths
- robustly parse included HTTP headers without splitting on blank lines inside JSON bodies
- add regression coverage for source files, byte-identical helper mirrors, the exact five-argument compatibility-wrapper contract, and the generated merged skill artifact

## Security impact

The helper no longer reads, copies, prints, or injects stored GitLab credentials. Authentication remains inside `glab`. Raw comment bodies are no longer echoed, control and bidirectional characters are rejected from printed file paths, API-controlled progress values are sanitized, and sensitive/debug header output is bounded and redacted.

## Validation

- `python3 -m unittest discover -s tests -v` — 19 passed
- fake-`glab` end-to-end subprocess test covers identity, versions, paginated diffs, split stdout/stderr HTTP validation details, ANSI injection on API-controlled success values, and successful `line_code` retry
- regression sabotage: the artifact/source regression test fails against `origin/main` on the original `"glab", "config", "get", "token"` helper pattern, then passes on this branch
- all three helper mirrors are byte-identical
- Python compile checks for all three mirrored helpers and Python 3.6 grammar checks for the helpers and test module
- `bash -n scripts/add-inline-comment.sh`
- included-response parser verified with multiple HTTP header blocks and pretty JSON containing blank lines
- extracted documentation example: Bash syntax valid and no unsupported `glab api --jq` flag
- live read-only `glab api --include` response parsing against GitLab.com
- live read-only GraphQL query through `glab api --header 'Content-Type: application/json' --input -`
- `npx --yes skills add . --list` — 48 skills discovered
- `bash scripts/build-claude-skill.sh --output <temporary artifact>`
  - exactly `gitlab-cli-skills/` and `gitlab-cli-skills/SKILL.md`
  - zero token-extraction/manual-header patterns in the merged skill
- `git diff --check`

## Compatibility / risk

- the helper remains compatible with its documented Python 3.6+ runtime
- bare `GITLAB_HOST` values are normalized to HTTPS by the legacy shell wrapper; non-default ports are rejected because `glab api --hostname` does not accept them
- the compatibility wrapper now rejects extra positional arguments so comment text must be passed as one quoted fifth argument, matching its documented contract
- no version bump, release, publication, or real MR comment write is included

Fixes #86

### CODE REVIEW

**Review by vince-winkintel — COMMENTED** [review]
Reviewed against #86. Verified locally: all 13 tests pass in a clean worktree, the three helper mirrors are byte-identical, `bash -n` passes on the wrapper, and the SKILL.md heredoc-into-pipe example is valid bash. One medium finding (unsanitized API-controlled terminal output on the success path, which is an explicit acceptance criterion of the issue) plus a few nits — see inline comments.

File: scripts/post-inline-comment.py [hunk 1]
```diff
@@ -214,32 +187,133 @@ def load_batch_file(path):
...
+
 
+def verify_glab_identity(host):
+    """Verify and display the selected GitLab host/account before write workflow calls."""
+    user = _api_get(host, "/user")
+    username = (user.get("username") or "").strip() if isinstance(user, dict) else ""
+    if not username:
+        raise GlabApiError("glab /user response did not include a non-empty username")
+    print(f"GitLab host: {get_hostname(host)}")
+    print(f"GitLab user: {username}")
```
Comment (vince-winkintel) [inline scripts/post-inline-comment.py:335]: **Medium — API-controlled terminal control sequences reach stdout on the success path.**

`username` is printed raw from the API response. The host is caller-influencible via `--host`/`GITLAB_HOST`, so a hostile or compromised self-hosted instance can inject ANSI escapes through this identity preflight. Confirmed by experiment: a response of `{"username": "bob\x1b[2J\x1b[Hpwned"}` puts raw escapes on the terminal.

The same class applies to `disc_id` (line 571, `... | disc_id: {disc_id}`) and the first 12 chars of `head_sha` (line 550), both also API-controlled.

Issue #86's acceptance criteria say "user/API-controlled terminal control sequences cannot reach progress or error output" — and the tests already assert this property for *error* output. Suggest routing these values through `sanitize_glab_error` (or a shared `sanitize_terminal()` helper) before printing, in all three helper mirrors.

File: scripts/post-inline-comment.py [hunk 2]
```diff
@@ -38,96 +38,69 @@
...
-    sys.exit(1)
-
-
-def _validate_token(token):
-    """Basic sanity check — token must look like a PAT (non-empty, no whitespace)."""
-    if not token or len(token) < 10 or re.search(r'\s', token):
-        print("ERROR: GITLAB_TOKEN appears invalid (too short or contains whitespace).", file=sys.stderr)
-        sys.exit(1)
-
+VALID_FILE_RE    = re.compile(r'^[^\x00-\x1f\x7f]+$')      # no terminal control characters
```
Comment (vince-winkintel) [inline scripts/post-inline-comment.py]: **Low — C1 control characters pass file-path validation.**

`^[^\x00-\x1f\x7f]+$` blocks C0 controls but admits C1 controls (U+0080–U+009F — e.g. U+009B is a single-byte CSI in some terminals) and bidi overrides like U+202E. `sanitize_glab_error` already strips C1; suggest matching that bar here, e.g. also excluding `\x80-\x9f`.

File: scripts/post-inline-comment.py [hunk 3]
```diff
@@ -214,32 +187,133 @@ def load_batch_file(path):
...
+        if char in "\n\t" or 0x20 <= ord(char) < 0x7f or ord(char) >= 0xa0
+    )
+    text = re.sub(
+        r'(?im)^(\s*>?\s*)(Authorization|PRIVATE-TOKEN|JOB-TOKEN|OAUTH-TOKEN)\s*:.*$',
+        lambda match: "%s%s: [REDACTED]" % (match.group(1), match.group(2)),
+        text,
+    )
+    text = re.sub(r'glpat-[A-Za-z0-9_\-]+', 'glpat-[REDACTED]', text)
+    text = re.sub(r'(?i)(PRIVATE-TOKEN|Authorization|Bearer|token)([=: ]+)(\S+)', r'\1\2[REDACTED]', text)
+    text = text.replace("\r", "\n").strip()
```
Comment (vince-winkintel) [inline scripts/post-inline-comment.py]: Nit: this `replace("\r", "\n")` is dead code — the character filter above already dropped `\r` (it keeps only `\n`, `\t`, and printable ranges). Worth removing, or reordering intentionally if the CR→LF normalization was meant to happen first.

File: scripts/post-inline-comment.py [hunk 4]
```diff
@@ -214,32 +187,133 @@ def load_batch_file(path):
...
+    if len(text) > limit:
+        text = text[:limit] + "...(truncated)"
+    return text
+
+
+def parse_included_response(output):
+    """Parse `glab api --include` output into (JSON body, response headers)."""
+    normalized = output.replace("\r\n", "\n")
+    header_text, sep, body_text = normalized.rpartition("\n\n")
```
Comment (vince-winkintel) [inline scripts/post-inline-comment.py]: Low: `rpartition("\n\n")` splits on the *last* blank line, which assumes the body never contains one. Safe today because glab emits compact JSON when stdout is piped, but a pretty-printed body with a blank line would mis-split. A comment documenting the assumption (or splitting on the first blank line after the final header block) would future-proof this.

File: tests/test_post_inline_comment_glab_api.py [hunk 5]
```diff
@@ -0,0 +1,381 @@
...
+class ArtifactRegressionTests(unittest.TestCase):
+    def test_source_and_merged_artifact_do_not_extract_plaintext_glab_tokens(self):
+        """Catches reintroducing the reported glab config token extraction snippet."""
+        forbidden_patterns = [
+            re.compile(r'glab\s+config\s+get\s+token'),
+            re.compile(r'["\']glab["\']\s*,\s*["\']config["\']\s*,\s*["\']get["\']\s*,\s*["\']token["\']'),
+            re.compile(r'PRIVATE-TOKEN["\']?\s*:\s*token'),
+            re.compile(r'glab\s+auth\s+token'),
+        ]
+        source_paths = [
```
Comment (vince-winkintel) [inline tests/test_post_inline_comment_glab_api.py:397]: Test-coverage suggestion: the functional tests only load `scripts/post-inline-comment.py`, and this forbidden-pattern check is the only guard over the other two mirrors — functional drift between them wouldn't be caught. A one-liner asserting the three helpers are byte-identical (e.g. comparing `read_bytes()`) would lock that in cheaply.

File: scripts/add-inline-comment.sh [hunk 6]
```diff
@@ -1,141 +1,35 @@
...
-    # Output JSON for programmatic use
-    echo "$BODY"
-else
-    echo "" >&2
-    echo -e "${RED}❌ Error: HTTP $HTTP_CODE${NC}" >&2
-    echo "$BODY" | jq . >&2
-    exit 1
-fi
+shift 4
+COMMENT_TEXT="$*"
```
Comment (vince-winkintel) [inline scripts/add-inline-comment.sh]: Nit: `shift 4; COMMENT_TEXT="$*"` silently joins any extra positional args into the comment body, where the old script ignored everything past `$5`. Reasonable, but it's an undocumented change from the stated 5-arg contract — worth a mention in the usage text or an explicit arg-count check.

File: scripts/post-inline-comment.py (hunk 1 above)
Comment (vince-winkintel) [reply scripts/post-inline-comment.py:335]: Addressed: added a shared terminal sanitizer and applied it to host, username, head SHA, discussion ID, and exception output in all three helper mirrors. The fake-glab end-to-end test now injects ANSI sequences into API-controlled success values and verifies no raw escape reaches stdout.

File: scripts/post-inline-comment.py (hunk 2 above)
Comment (vince-winkintel) [reply scripts/post-inline-comment.py]: Addressed: file-path validation now rejects the full C1 range plus Unicode bidi controls (including U+202E), uses full-match semantics so trailing newlines cannot slip through, and has focused regression coverage.

File: scripts/post-inline-comment.py (hunk 3 above)
Comment (vince-winkintel) [reply scripts/post-inline-comment.py]: Addressed: CRLF/CR normalization now happens before terminal-character filtering in the shared sanitizer, and a regression test verifies the intended multiline result.

File: scripts/post-inline-comment.py (hunk 4 above)
Comment (vince-winkintel) [reply scripts/post-inline-comment.py]: Addressed: included responses are parsed from the final HTTP status block and split at that block's first blank line, so blank lines in pretty-printed JSON bodies remain intact. Verified with pretty JSON and multiple HTTP response blocks.

File: tests/test_post_inline_comment_glab_api.py (hunk 5 above)
Comment (vince-winkintel) [reply tests/test_post_inline_comment_glab_api.py:397]: Addressed: added a regression test that reads all three helper copies as bytes and requires them to be identical.

File: scripts/add-inline-comment.sh (hunk 6 above)
Comment (vince-winkintel) [reply scripts/add-inline-comment.sh]: Addressed: documented that comment text is one quoted fifth argument, enforced exactly five positional arguments, and added a test proving extra arguments are rejected.

**Review by vince-winkintel — COMMENTED** [review]
Vince: Resolved
